amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java

```diff
@@ -473,14 +474,180 @@ public void testCheckAndCreateLeasesForShardsWithEmptyShardList() throws Excepti
+        // lease for shard-0 when reading from AT_TIMESTAMP.
+        final Set<String> missingLeaseKeys = new HashSet<>(Arrays.asList("shardId-0", "shardId-6", "shardId-8"));
+        final List<Shard> shardsWithLeases = shards.stream()
+                .filter(s -> !missingLeaseKeys.contains(s.shardId())).collect(Collectors.toList());
+        final List<Lease> existingLeases = createLeasesFromShards(shardsWithLeases, ExtendedSequenceNumber.AT_TIMESTAMP, LEASE_OWNER);
+
+        final Set<String> expectedLeaseKeysToCreate = new HashSet<>(Arrays.asList("shardId-0"));
+        testCheckAndCreateLeaseForShardsIfMissing(shards, INITIAL_POSITION_AT_TIMESTAMP, expectedLeaseKeysToCreate, existingLeases);
+    }
+
+    /*
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Missing leases: (0, 6, 8)
+     * Initial position: AT_TIMESTAMP(200)
+     * Leases to create: (0)
+     */
     @Test
-    public void testCheckAndCreateLeasesForNewShardsAtTrimHorizon() throws Exception {
-        testCheckAndCreateLeaseForShardsIfMissing(constructShardListForGraphA(), INITIAL_POSITION_TRIM_HORIZON);
+    public void testCheckAndCreateLeasesForNewShardsAtTimestampWithPartialLeaseTable2() throws Exception {
+        final List<Shard> shards = constructShardListForGraphA();
+        final InitialPositionInStreamExtended initialPosition = InitialPositionInStreamExtended
+                .newInitialPositionAtTimestamp(new Date(200L));
+        // Leases for shard-0 and its descendants (shard-6, and shard-8) are missing. Expect lease sync to recover the
+        // lease for shard-0 when reading from AT_TIMESTAMP.
+        final Set<String> missingLeaseKeys = new HashSet<>(Arrays.asList("shardId-0", "shardId-6", "shardId-8"));
+        final List<Shard> shardsWithLeases = shards.stream()
+                .filter(s -> !missingLeaseKeys.contains(s.shardId())).collect(Collectors.toList());
+        final List<Lease> existingLeases = createLeasesFromShards(shardsWithLeases, ExtendedSequenceNumber.AT_TIMESTAMP, LEASE_OWNER);
+
+        final Set<String> expectedLeaseKeysToCreate = new HashSet<>(Arrays.asList("shardId-0"));
+        testCheckAndCreateLeaseForShardsIfMissing(shards, initialPosition, expectedLeaseKeysToCreate, existingLeases);
     }
 
+    /*
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Missing leases: (0, 6, 8)
+     * Initial position: LATEST
+     * Leases to create: (0)
```

Comment:
Can you add the following cases as well?
LATEST
1. If the missing leases are 0,1,6,8 or 0,1,6,7,8, then we  need to create lease only for 6.
2. If the missing leases are 0,1,2,3,6,7 then we create leases for 6 and 7
3. If the missing leases are 1,2 or 1,2,6 or 1,2,6,7, we need to create leases for 0 and 3.

TRIM/TIMESTAMP
1. If the missing leases are 0,1,6,8 or 0,1,6,7,8, then we need to create lease only for 0,1.
2. If the missing leases are 0,1,2,3,6,7 then we create leases for 0,1,2,3
3. If the missing leases are 1,2 or 1,2,6 or 1,2,6,7, we need to create leases for 0 and 3.

Reply:
LATEST:
1. Creating a lease only for 6 and not for 0,1 is covered in testDetermineNewLeasesToCreateSplitMergeLatestA_PartialHashRange1()
2. Wouldn't we create only a lease for 8 in this case?
3.

Reply:
Discussed offline, these cases are covered